### PR TITLE
Segment contact procedure

### PR DESCRIPTION
**Pull request by iBatistic**
In progress...

### CODE REVIEW

**Comment by philipcardiff** [comment]
This looks great. I will go through the changes and let you know of any edits before merging.

File: src/solids4FoamModels/numerics/newGGIInterpolation/newGGIInterpolationPolygonIntersection.C [hunk 1]
```diff
@@ -333,6 +341,264 @@ scalar newGGIInterpolation<MasterPatch, SlavePatch>::area2D
...
+    const vector& slaveFaceNormal = slavePatch_.faceNormals()[faceSi];
+    const vector& masterFaceNormal = masterPatch_.faceNormals()[faceMi];
+    const pointField& slavePatchPoints = slavePatch_.localPoints();
+    const pointField& masterPatchPoints = masterPatch_.localPoints();
+
+    const vector& slaveFaceCentre =
+        slavePatch_[faceSi].centre(slavePatchPoints);
+
+    // Master face center is offset for ortho base c.s.
+    const vector& orthoBaseOffset =
```
Comment (philipcardiff) [inline src/solids4FoamModels/numerics/newGGIInterpolation/newGGIInterpolationPolygonIntersection.C:365]: Same as above. This should not be a reference as `face::centre(const pointField&) const` returns a copy of a point (defined in face.H as far as I can see), i.e. switch this to `const vector orthoBaseOffset =`.

**Review by philipcardiff — COMMENTED** [review]
See my specific comments above.

**Comment by iBatistic** [comment]
> See my specific comments above/below.

@philipcardiff, thanks for the comments. I still have a few things to add to finish code.

> By the way, merging solidContact and solidSegmentContact is very nice. We can probably change the default behaviour to solidSegmentContact once we check the tutorials still work.

I'm trying to add things with as few modifications as possible.   It would be nice also to create a single boundary condition that can work implicitly and explicitly.

**Comment by philipcardiff** [comment]
HI @iBatistic, is everything now resolved? If so, I will merge this.

**Comment by iBatistic** [comment]
@philipcardiff It's not over yet, I still have to add the code in `solidContact` for the tangential force component.

**Comment by philipcardiff** [comment]
> @philipcardiff It's not over yet. Your comments are resolved, but I need to add code to `solidContact` for the tangential force component.

No problem!

**Comment by iBatistic** [comment]
@philipcardiff, I still have to test everything, so don't do the merge yet.

**Comment by philipcardiff** [comment]
@iBatistic: this all looks fine to me.

Is there a test case I should try using the new contact approach (maybe any/everyone is suitable)?
Based on your thesis, the new approach should be the default approach (maybe we can only do this for FE41, for now).
Is adding `integralNormalGap yes;` the current correct way to use the new approach?

Philip

**Comment by iBatistic** [comment]
@philipcardiff I just realized that the code can't be compiled with other versions, I'll add the required changes tomorrow

**Comment by iBatistic** [comment]
> @philipcardiff I just realized that the code can't be compiled with other versions, I'll add the required changes tomorrow

Done in the latest commit.

**Comment by philipcardiff** [comment]
> Before merging, I would add one more change; a warning to the user if `integralNormalGap` is set, but the user is not on fe.
> 
> **Edit**: I will not add this because then we also need to add a warning for other switches like `useNewPointDistanceMethod` and `usePrevCandidateMasterNeighbors`.

This is an interesting question. I agree that maintaining consistency is a good thing; having said that, it seems like a pity that a superior method may not be used in FE41 (and users may not realise it is available). So, another approach is to enable it and give a warning for other versions; that way, users of other versions will be made aware of the other (better) approach. That may motivate people to try the FE41 or even consider helping port the code to OF.com/OF.org. What are your thoughts?

**Comment by philipcardiff** [comment]
Merged now, but we can still change the default behaviour if needed, pending the outcome of our discussion above.

**Comment by iBatistic** [comment]
> This is an interesting question. I agree that maintaining consistency is a good thing; having said that, it seems like a pity that a superior method may not be used in FE41 (and users may not realise it is available). So, another approach is to enable it and give a warning for other versions; that way, users of other versions will be made aware of the other (better) approach. That may motivate people to try the FE41 or even consider helping port the code to OF.com/OF.org. What are your thoughts?

I agree, I think that your idea is a better approach. You can change the default behaviour in some of your next commits to avoid opening a new PR.
Also, I think it would be useful to create documentation for the contact boundary condition to give people a better idea of how things work. I will start working on documentation and hope to make PR for it next week.

**Pull request merged**